pkg/cache/cache_test.go

```diff
@@ -22,40 +22,96 @@ import (
+const testNamespaceTwo = "test-namespace-2"
+
+func createPod(name, namespace string, restartPolicy kcorev1.RestartPolicy) runtime.Object {
+	three := int64(3)
+	pod := &kcorev1.Pod{
+		ObjectMeta: kmetav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+			Labels: map[string]string{
+				"test-label": name,
+			},
+		},
+		Spec: kcorev1.PodSpec{
+			Containers:            []kcorev1.Container{{Name: "nginx", Image: "nginx"}},
+			RestartPolicy:         restartPolicy,
+			ActiveDeadlineSeconds: &three,
+		},
+	}
+	cl, err := client.New(cfg, client.Options{})
+	Expect(err).NotTo(HaveOccurred())
+	err = cl.Create(context.Background(), pod)
+	Expect(err).NotTo(HaveOccurred())
+	return pod
+}
+
+func deletePod(pod runtime.Object) {
+	cl, err := client.New(cfg, client.Options{})
+	Expect(err).NotTo(HaveOccurred())
+	err = cl.Delete(context.Background(), pod)
+	Expect(err).NotTo(HaveOccurred())
+}
+
 var _ = Describe("Informer Cache", func() {
-	var stop chan struct{}
+
+	var (
+		informerCache cache.Cache
+		stop          chan struct{}
+		knownPod1     runtime.Object
+		knownPod2     runtime.Object
+		knownPod3     runtime.Object
+	)
 
 	BeforeEach(func() {
 		stop = make(chan struct{})
 		Expect(cfg).NotTo(BeNil())
+
+		By("creating three pods")
+		knownPod1 = createPod("test-pod-1", testNamespaceOne, kcorev1.RestartPolicyNever)
+		knownPod2 = createPod("test-pod-2", testNamespaceTwo, kcorev1.RestartPolicyAlways)
+		knownPod3 = createPod("test-pod-3", testNamespaceTwo, kcorev1.RestartPolicyOnFailure)
+
+		By("creating the informer cache")
+		var err error
+		informerCache, err = cache.New(cfg, cache.Options{})
+		Expect(err).NotTo(HaveOccurred())
+		By("running the cache and waiting for it to sync")
+		go func() {
+			defer GinkgoRecover()
+			Expect(informerCache.Start(stop)).ToNot(HaveOccurred())
+		}()
+		Expect(informerCache.WaitForCacheSync(stop)).NotTo(BeFalse())
 	})
+
 	AfterEach(func() {
+		By("cleaning up created pods")
+		deletePod(knownPod1)
+		deletePod(knownPod2)
+		deletePod(knownPod3)
+
 		close(stop)
 	})
 
 	Describe("as a Reader", func() {
 		It("should be able to list objects that haven't been watched previously", func() {
-			By("Creating the cache")
-			reader, err := cache.New(cfg, cache.Options{})
-			Expect(err).NotTo(HaveOccurred())
-
-			By("running the cache and waiting for it to sync")
-			go func() {
-				defer GinkgoRecover()
-				Expect(reader.Start(stop)).ToNot(HaveOccurred())
-			}()
-			Expect(reader.WaitForCacheSync(stop)).NotTo(BeFalse())
-
-			By("Listing all services in the cluster")
+			By("listing all services in the cluster")
 			listObj := &kcorev1.ServiceList{}
-			Expect(reader.List(context.Background(), nil, listObj)).NotTo(HaveOccurred())
+			Expect(informerCache.List(context.Background(), nil, listObj)).NotTo(HaveOccurred())
 
-			By("Verifying that the returned list contains the Kubernetes service")
+			By("verifying that the returned list contains the Kubernetes service")
```

Comment:
where does the service come from?  I don't see it created anywhere.

Reply:
Updated "NB: kubernetes default service is automatically created in testenv. " on line 117.